In [56]:
import numpy as np
import string
from collections import defaultdict

In [2]:
dataset_dir = 'Dataset/'

In [3]:
with open(dataset_dir+'all_tags_universal.npy', 'rb') as f:
    all_tags_universal = np.load(f)
all_tags_universal.shape

(56052,)

In [5]:
with open(dataset_dir+'all_questions_final.npy', 'rb') as f:
    all_ques = np.load(f)
all_ques.shape

(56052,)

In [51]:
with open(dataset_dir+'tagset_universal.npy', 'rb') as f:
    tagset_uni = np.load(f)
tagset_uni.shape

(12,)

In [64]:
with open(dataset_dir+'wordIndex.npz', 'rb') as f:
    wordIndex = np.load(f)['a']

array(['"7"', '"adopted"', '"alley', ..., 'zozo', 'zu', 'zucchini'],
      dtype='<U20')

In [65]:
wordIndex = wordIndex.tolist()

DOING OVERLAP BASED ON POS TAGS

In [21]:
primaryDict = defaultdict((int)) # tuple for ints of any length

for tag in all_tags_universal:
    for val in range(1,len(tag)+1):
        primaryDict[tag[0:val]] += 1

In [47]:
new = []
ofLengthAtLeast = 3

for key in primaryDict.keys():
    if len(key) > ofLengthAtLeast: # long enough to prompt duplication over multiple values
        new.append((key, (primaryDict[key])))

In [48]:
newMax = sorted(new, key = lambda x: x[1], reverse=True)

In [49]:
newMax

[((7, 4, 1, 4), 4577),
 ((7, 4, 1, 4, 3), 4375),
 ((7, 3, 6, 4), 3983),
 ((11, 3, 6, 4), 3069),
 ((7, 4, 3, 6), 2309),
 ((3, 6, 4, 3), 2277),
 ((7, 4, 1, 4, 3, 1), 2094),
 ((11, 9, 3, 6), 2023),
 ((11, 3, 7, 3), 1824),
 ((11, 9, 3, 6, 4), 1783),
 ((11, 3, 6, 4, 3), 1666),
 ((7, 4, 1, 4, 3, 6), 1524),
 ((7, 3, 6, 4, 1), 1457),
 ((3, 7, 3, 6), 1407),
 ((3, 6, 6, 4), 1288),
 ((7, 3, 7, 3), 1281),
 ((3, 6, 4, 1), 1240),
 ((7, 3, 6, 4, 1, 6), 1149),
 ((7, 4, 3, 1), 1127),
 ((6, 3, 6, 4), 1087),
 ((3, 7, 3, 6, 4), 1078),
 ((3, 6, 4, 4), 1043),
 ((7, 4, 3, 6, 4), 1000),
 ((3, 1, 6, 4), 950),
 ((7, 3, 6, 4, 1, 6, 4), 946),
 ((7, 3, 6, 4, 3), 878),
 ((11, 9, 4, 3), 851),
 ((3, 6, 4, 1, 6), 785),
 ((7, 4, 3, 7), 773),
 ((6, 3, 7, 3), 771),
 ((11, 11, 3, 7), 756),
 ((11, 9, 3, 1), 716),
 ((3, 6, 9, 4), 710),
 ((11, 11, 3, 7, 3), 708),
 ((7, 3, 6, 4, 4), 691),
 ((3, 6, 4, 1, 6, 4), 672),
 ((11, 11, 3, 6), 665),
 ((3, 7, 3, 1), 647),
 ((11, 9, 3, 1, 4), 611),
 ((11, 3, 7, 3, 6), 609),
 ((11, 11, 3,

In [50]:
print (newMax[0])

((7, 4, 1, 4), 4577)


In [53]:
for val in newMax[0][0]:
    print (tagset_uni[val])

PRON
NOUN
ADP
NOUN


# Now we party, cuz 4577 is a crowd!

CHECKING FOR OVERLAP BASED ON ACTUAL WORDS

In [86]:
# converting to the numerical data becasuse redundancy

newList = []
for sent in all_ques:
    temp = []
    for word in sent:
        temp.append(wordIndex.index(word))
    newList.append(tuple(temp))

In [71]:
with open('all_questions_final_num.npy', 'wb') as f:
    np.save(f, newList)

One time save operation

In [87]:
newList

[(4821, 9213, 1726, 119, 7043),
 (4535, 5372, 9945, 9213, 6934),
 (4259, 9257, 1141, 868, 6935, 6185),
 (2801, 9213, 8835, 3120, 9717, 5598, 9213, 6643),
 (10033, 4821, 4285, 9119, 4353),
 (10080, 4821, 406, 8424, 8946),
 (10080, 4821, 8070, 3249),
 (4535, 5916, 2669, 8070, 3886, 6370, 9335, 826, 4661, 9257, 3462),
 (10033, 5903, 4821, 9257, 7858, 3743),
 (10033, 5903, 9945, 9257, 3743),
 (10080, 4821, 9257, 5491, 4661, 4203),
 (10080, 4821, 9257, 5491, 4661, 4203),
 (10080, 4821, 9201, 5491, 4661, 4203),
 (10080, 4821, 4285, 4661, 2383),
 (504, 9264, 307, 2446, 2263),
 (10080, 504, 9237, 1729, 6209, 9213, 3372),
 (504, 9237, 10097, 2263),
 (10023, 9213, 3912, 119, 5321, 984, 234, 6168, 9213, 2263),
 (504, 9264, 2263, 7746, 9264, 1729),
 (10033, 5016, 6168, 9121, 4821, 9257),
 (10033, 2801, 9213, 166, 9121, 2785),
 (10033, 9640, 6168, 166, 9945, 9257, 9121, 6576),
 (10033, 9945, 9213, 3951, 6168, 9257, 9121),
 (2785, 9237, 6554, 9880, 3642, 119, 1655),
 (10069, 4821, 9213, 2618, 6168, 

In [103]:
secondaryDict = defaultdict((int)) # tuple for ints of any length

for sent in newList:
    for val in range(1,len(sent)+1):
        secondaryDict[sent[0:val]] += 1

In [104]:
secondaryDict

defaultdict(int,
            {(4821,): 8893,
             (4821, 9213): 1336,
             (4821, 9213, 1726): 14,
             (4821, 9213, 1726, 119): 5,
             (4821, 9213, 1726, 119, 7043): 1,
             (4535,): 7148,
             (4535, 5372): 1117,
             (4535, 5372, 9945): 79,
             (4535, 5372, 9945, 9213): 55,
             (4535, 5372, 9945, 9213, 6934): 1,
             (4259,): 73,
             (4259, 9257): 14,
             (4259, 9257, 1141): 1,
             (4259, 9257, 1141, 868): 1,
             (4259, 9257, 1141, 868, 6935): 1,
             (4259, 9257, 1141, 868, 6935, 6185): 1,
             (2801,): 1210,
             (2801, 9213): 376,
             (2801, 9213, 8835): 1,
             (2801, 9213, 8835, 3120): 1,
             (2801, 9213, 8835, 3120, 9717): 1,
             (2801, 9213, 8835, 3120, 9717, 5598): 1,
             (2801, 9213, 8835, 3120, 9717, 5598, 9213): 1,
             (2801, 9213, 8835, 3120, 9717, 5598, 9213, 6643): 1,
        

In [111]:
newWord = []
ofLengthAtLeast = 1

for key in secondaryDict.keys():
    if len(key) > ofLengthAtLeast: # long enough to prompt duplication over multiple values
        newWord.append((key, (secondaryDict[key])))

In [112]:
newWordMax = sorted(newWord, key = lambda x: x[1], reverse=True)

In [139]:
total = sum(b for a,b in newWordMax)
total

285535

In [143]:
sum([b/total for a,b in newWordMax if a[0]==10033])

0.31300541089536615

[((10033, 4821), 0.015984029978811706),
 ((10033, 5016), 0.011094962088710665),
 ((10033, 5016, 6168), 0.011073948902936592),
 ((4821, 9257), 0.0108463060570508),
 ((4821, 9201), 0.00975011819916998),
 ((10033, 4821, 9213), 0.007011399653282435),
 ((10053, 4821), 0.006524594182849738),
 ((4535, 6194), 0.005827656854676309),
 ((10080, 4821), 0.005410895336823857),
 ((10033, 504), 0.0053338469889855885),
 ((4535, 6194, 4821), 0.005039662388148563),
 ((4821, 9213), 0.004678936032360306),
 ((10069, 4821), 0.0043952580244103175),
 ((10053, 4821, 9257), 0.0041255888069763776),
 ((4535, 5372), 0.003911954751606633),
 ((4821, 9257, 119), 0.003740347067785035),
 ((4821, 9201, 119), 0.0037368448701560228),
 ((10033, 9640), 0.003379620711996778),
 ((4535, 5916), 0.003362109723851717),
 ((10033, 9640, 6168), 0.003344598735706656),
 ((2669, 10280), 0.0031975064352881433),
 ((4535, 5530), 0.0031379690755949357),
 ((504, 9237), 0.0031204580874498746),
 ((504, 9239), 0.003008387763321484),
 ((2785, 10

In [126]:
for val in newWordMax[0:10]:
    print ("**********")
    test = val[0]
    for val in test:
        print (wordIndex[val])

**********
what
is
**********
what
kind
**********
what
kind
of
**********
is
this
**********
is
that
**********
what
is
the
**********
where
is
**********
how
old
**********
why
is
**********
what
are


Only 369 of "what is the name" // AND // Only 4564 of "what is"